:sparkles:feat: config params, nonce management, cli

ibourn opened the pull request:
- Refactor `constants` folder:
  - replace  config js files with json in `constants/`
  - refactor config object: remove `tokenParams`, add `usedConfigs` and `activeConfig`
  - move fork ports in `.env` file and adapt hardhat config file
- CLI:
  - add hardhat tasks to manage config params
  - begin test for these tasks
- Manage multiple nonces for user actions to handle non sequential actions:
  - refactor `nonceRecord` to:
    - add `bridgeAddress` in order to use the last deployed contract
    - an array of current nonces for a user address
- Refactor helpers to fit above changes
- Clean hepers
- Update COMMANDS & README files

- [x] #20 
- [x] #27 
- [x] #29

ibourn committed:
:perf:feat: User actions - add multi nonce management

- Install proper-lockfile package
- Refactor write/read func for nonce in fileHelper
- Add file lock mechanism to manage concurent use of the json file
- Add user address entrie in the nonceRecord json
- Update script 11 and 12 accordingly

ibourn committed:
:sparkles:feat: draft config setup cli

- Move config params from js file to json format
- Move networkParam, tokenParams, usedNetworks, usedTokens, forkPorts
- Add file helper to read networks and tokens config
- Add a label 'activeConfig' to switch between preconfigured config
- Begin cli to select a usedConfig (add, select active, list)

ibourn committed:
:white_check_mark:test: begin tasks test & add cli feat

- add use of real deployed token
- add config helper
- add helper to switch active config
- adapt script to new features
- begin tasks tests

ibourn committed:
:recycle:refactor: config helper - add tests

- move params object logic from file to config helper
- refacto config helpers
- add tests for config task

ibourn committed:
:recycle:refactor: clean new updates

- refacto scripts & tasks after adding config cli

ibourn committed:
:pencil2:refactor: update COMMANDS & README files

ibourn committed:
:bug:fix: bug with nonces & remove js data file

- fix many minor typos across files
- replace simulationParams.js with simulationConfig.json
- adapt scripts to the new file
- add helpers to manage it
- add bridge address to nonceRecord to use the last contract deployed
- adapt userAction scripts & fileHelper to the new nonce object
- move ports from json file to .env and adapt task & hh config

ibourn committed:
:pencil2:refactor: renaming allfeat to harmonie (AFT->HMY)

ibourn committed:
:recycle:perf: refactor storage ref in Bridge and Relayer

 - replace storage address by private immutable Storage instance
 - replace all cast from storage address to instance with the new variable
 - gas saving: average: 2-6K

ArnaudSene commented on hardhat/hardhat.config.js:
```diff
@@ -65,41 +67,41 @@ module.exports = {
+    // allfeat: {
+    //   url: `https://harmonie-endpoint-02.allfeat.io`,
+    //   accounts: [deployerPvtKey, ...usersPvtKeys],
+    // },
```
Is there a good reason to keep this code commented?

ibourn replied:
no, mainnet will be added when launched

ArnaudSene commented on hardhat/helpers/configHelper.js:
```diff
@@ -1,251 +1,450 @@
+  if (!shouldExist && obj[key]) {
+    throw new Error(`'${key}' already exists.`);
+  }
+}
```
`function assertKeyExistence(obj, key, shouldExist = true): void {
  const exists = key in obj;

  if (shouldExist && !exists) {
    throw new Error(`'${key}' does not exist.`);
  }

  if (!shouldExist && exists) {
    throw new Error(`'${key}' already exists.`);
  }
}
`

Using `in` operator to check key is more robust because if key is `undefined` or `null` result will be `undefined` and not a boolean.
I propose a function name `assertKeyExistence` to be more relevant

ArnaudSene commented on hardhat/helpers/configHelper.js:
```diff
@@ -1,251 +1,450 @@
+ * @param {any} defaultValue The default value to set the key to
+ */
+function ensureExist(obj, key, defaultValue) {
+  if (!obj[key]) {
```
`(!(key in obj))`

Use `in` to check and return a boolean

ArnaudSene commented on hardhat/helpers/configHelper.js:
```diff
@@ -1,251 +1,450 @@
+ * @param {Object} token The token to check
+ */
+function ensureTokenExists(registry, token) {
+  if (!registry[token.name]) {
```
`(!(token.name in registry))`
Use `in` operator

ArnaudSene commented on hardhat/helpers/configHelper.js:
```diff
@@ -1,251 +1,450 @@
+ * @throws {Error} If the address is provided for mocked tokens
+ */
+function addDeployedToken(networkName, tokenName, tokenSymbol, tokenAddress) {
+  if (!configParams.networkParams || !configParams.networkParams[networkName]) {
```
`(!configParams.networkParams?.[networkName]) `

Shorter

ArnaudSene commented on hardhat/helpers/configHelper.js:
```diff
@@ -1,251 +1,450 @@
+  try {
+    if (label !== "usedNetworks" && label !== "usedTokens") {
+      throw new Error(`Label '${label}' not recognized.`);
+    }
```
`
const validLabels = ["usedNetworks", "usedTokens"];

if (!validLabels.includes(label)) {
  throw new Error(`Label '${label}' not recognized.`);
}
`

Suggestion

ibourn replied:
yes it is more explicit

ArnaudSene commented on hardhat/helpers/functionHelpers.js:
```diff
@@ -294,10 +318,50 @@ const convertParamsStringToArray = (argsString) => {
+function buildRpcUrl(networkName) {
+  // insert '_' before exact words: Fork or Local and convert to uppercase
+  const envName =
+    networkName == "geth" ? "PORT_GETH" : networkNameToPortName(networkName);
```
Don't you have a better alternative to avoid hard coding?

ibourn replied:
Modification done, but as you use "geth", i need to keep this key word hardcoded.

ArnaudSene commented on hardhat/scripts/03_setDeployedToken.js:
```diff
@@ -0,0 +1,94 @@
+
+  const options = process.env.MOCKED_TOKEN_OPTION;
+  const tokenInfo = options.split(",");
+  if (tokenInfo.length != 3) {
```
`if (tokenInfo.length !== 3) {`

ibourn replied:
forgotten

ArnaudSene commented on hardhat/scripts/03_setDeployedToken.js:
```diff
@@ -0,0 +1,94 @@
+  let tokenToSetObject = deployedTokens.find(
+    (element) => element.name === tokenToSet
+  );
+  if (!tokenToSetObject.address || tokenInfo[2] != tokenToSetObject.address) {
```
`if (!tokenToSetObject.address || tokenInfo[2] !== tokenToSetObject.address) {`

ibourn replied:
yes only ascii char in the address string

ArnaudSene commented on hardhat/scripts/12_userAction_depositFees.js:
```diff
@@ -2,16 +2,16 @@ const hre = require("hardhat");
 const { toStyle, display } = require("../helpers/loggingHelper");
 const { getContext } = require("../helpers/contextHelper");
-const { getNetworkNameByChainId } = require("../helpers/configHelper");
+// const { getNetworkNameByChainId } = require("../helpers/configHelper");
```
Should the commented code be removed?

ArnaudSene commented on hardhat/tasks/start-node.js:
```diff
@@ -1,22 +1,39 @@
       throw new Error(
         `Network ${taskArgs.networkToFork} is not configured correctly.`
       );
     }
```
`
if (!networkConfig?.url) {
  throw new Error(`Network '${taskArgs.networkToFork}' is not configured correctly.`);
}
`

ArnaudSene commented on hardhat/tasks/start-node.js:
```diff
@@ -1,22 +1,39 @@
+          taskArgs.networkToFork + "Fork"
+        } is not configured correctly.`
+      );
+    }
```
`
if (!forkNetworkConfig?.url) {
  throw new Error(`Fork network '${taskArgs.networkToFork}Fork' is not configured correctly.`);
}
`

ArnaudSene commented on hardhat/test/tasks/func-mintBridgedToken.js:
```diff
@@ -0,0 +1,113 @@
+
+//     console.error.restore();
+//   });
+// });
```
Why this code  is commented? Should be removed?

ibourn replied:
In progress, not working yet (I’m trying stub in js) but yes it can be removed.

ibourn committed:
:recycle:refactor following review

- fix many typos
- improve js syntax as suggested in many scripts
- removerd commented test file and unused comments
- refactor buildRpcUrl to avoid hardcoding
